base_layer/wallet/src/output_manager_service/service.rs

```diff
@@ -748,6 +759,110 @@ where
     }
 
+    pub async fn create_coin_split(
+        &mut self,
+        amount_per_split: MicroTari,
+        split_count: usize,
+        fee_per_gram: MicroTari,
+        lock_height: Option<u64>,
+    ) -> Result<(u64, Transaction, MicroTari, MicroTari), OutputManagerError>
+    {
+        trace!(
+            target: LOG_TARGET,
+            "Select UTXOs and estimate coin split transaction fee."
+        );
+        let total_split_amount = amount_per_split * split_count as u64;
+        let mut estimate_input_count: usize = 1;
+        let mut estimate_output_count = split_count;
+        let mut inputs = Vec::<UnblindedOutput>::new();
+        let mut estimated_fee = MicroTari::default();
+        let mut utxo_total = MicroTari::default();
+        loop {
+            estimated_fee = Fee::calculate(fee_per_gram, 1, estimate_input_count, estimate_output_count);
+            let total_amount = total_split_amount + estimated_fee;
+            inputs = self
+                .select_outputs(total_amount, 0.into(), UTXOSelectionStrategy::MaturityThenSmallest)
+                .await?;
+            // Check if estimated inputs correct
+            let input_count = inputs.len();
+            if input_count != estimate_input_count {
+                estimate_input_count = input_count;
+                continue;
+            }
+            // Check if estimated outputs correct
+            utxo_total = inputs.iter().fold(MicroTari::from(0), |acc, x| acc + x.value);
+            let output_count = if utxo_total >= total_amount {
+                split_count + 1
+            } else {
+                split_count
+            }; // Might require a change output
+            if output_count != estimate_output_count {
+                estimate_output_count = output_count;
+                continue;
+            }
+            // Check amounts
+            if utxo_total >= total_amount {
+                break;
+            }
+        }
+        let output_count = estimate_output_count;
+
+        trace!(target: LOG_TARGET, "Construct coin split transaction.");
+        let offset = PrivateKey::random(&mut OsRng);
+        let nonce = PrivateKey::random(&mut OsRng);
+        let mut builder = SenderTransactionProtocol::builder(0);
+        builder
+            .with_lock_height(lock_height.unwrap_or(0))
+            .with_fee_per_gram(fee_per_gram)
+            .with_offset(offset.clone())
+            .with_private_nonce(nonce.clone());
+        trace!(target: LOG_TARGET, "Add inputs to coin split transaction.");
+        for uo in inputs.iter() {
+            builder.with_input(
+                uo.as_transaction_input(&self.factories.commitment, uo.clone().features),
+                uo.clone(),
+            );
+        }
+        trace!(target: LOG_TARGET, "Add outputs to coin split transaction.");
+        let mut outputs = Vec::with_capacity(output_count);
+        let change_output = (utxo_total - estimated_fee) - total_split_amount;
```

Comment:
It shouldn't be possible, but potential underflow. Consider using `checked_sub`

Reply:
Change applied.